[SPARK-25299] Implement default version of the API for shuffle writes

ifilonenko opened the pull request:
## What changes were proposed in this pull request?

Implement default version of the API that would be used across all shuffle writers, writes to local disk.

Shuffle Writes [2/6] [3/6] [5/6]

## How was this patch tested?

Compiled and unit tests

ifilonenko committed:
musings without unit tests

ifilonenko commented:
@mccheah @yifeih for initial opinions

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShufflePartitionWriter.java:
```diff
@@ -0,0 +1,52 @@
+
+  @Override
+  public OutputStream openStream() throws IOException {
+    return stream;
```
Hmm, looking at this more carefully I don't think this is going to have the `transferTo` behavior that we expect. The reason for this is that `Channels.newChannel` knows how to make a `FileChannel` from a `FileOutputStream` specifically - it checks `instanceOf`. But here, `Channels.newChannel` would be passed a `DefaultShuffleBlockOutputStream`, which is not a `FileOutputStream`.

ifilonenko replied:
True, in its current form, it would return a `WritableByteChannelImpl`

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleWriteSupport.java:
```diff
@@ -0,0 +1,50 @@
+  private final ShuffleWriteMetricsReporter metrics;
+
+  public DefaultShuffleWriteSupport(
+          SparkConf sparkConf,
```
Indentation should be 4 spaces from `public`.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShufflePartitionWriter.java:
```diff
@@ -0,0 +1,52 @@
+
+import java.io.*;
+
+public class DefaultShufflePartitionWriter implements ShufflePartitionWriter {
```
This is rare, but in this case I think it makes more sense for this to be a static inner class of `DefaultMapOutputWriter` - the interactions between the length stored here and usage in `DefaultMapOutputWriter#commitAllPartitions` would be clearer this way. But I'd like a second opinion from @yifeih.

ifilonenko replied:
Yeah, I am okay with either. I thought this was more isolated.

yifeih replied:
I don't have a strong preference either way, although I did find myself needed to go to the other class to understand this one, so placing it together could help.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,171 @@
+      throw new IOException(
+              String.format(
+                      "Failed to create empty shuffle file at %s.", outputFile.getAbsolutePath()));
+    }
```
Do we have to clean up as part of `committing` as well?

ifilonenko replied:
Yes

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,171 @@
+
+  private void cleanUp() throws IOException {
+    if (outputBufferedFileStream != null) {
+      outputBufferedFileStream.flush();
```
Closing an underlying output stream I _believe_ implies flushing it, but double check this.

ifilonenko replied:
It does, very true. `close()` I believe always calls `flush()`

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,171 @@
+  @Override
+  public void abort(Throwable error) throws IOException {
+    cleanUp();
+    if (!outputTempFile.delete()) {
```
Check both here and in the next clause that the `outputTempFile` `exists` first. No point logging a warning trying to delete a file that never existed.

ifilonenko replied:
Good catch

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,171 @@
+
+  @Override
+  public void abort(Throwable error) throws IOException {
+    cleanUp();
```
When aborting, we don't want `cleanup` to throw an exception that halts the rest of this method - we want to always be able to attempt to delete the files.

mccheah reviewed (changes requested):
Initial review. But the behavior here is tricky - I think we'd get clearer validation that we're doing things correctly if we integrated this into one of the shuffle writers and ran the unit tests.

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleBlockOutputStream.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.io.OutputStream;
+
+public abstract class DefaultShuffleBlockOutputStream extends OutputStream {
```
Why can't it just be a class with the implementation contents instead of an abstract class? Will there ever need to be other implementations of this?

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,171 @@
+      ts = new TimeTrackingOutputStream(metrics, outputFileStream);
+    }
+    if (outputBufferedFileStream == null) {
+      outputBufferedFileStream = new BufferedOutputStream(ts, bufferSize);
```
Do you always want to use a BufferedOutputStream here? `BypassMergeSortShuffleWriter` doesn't use it at all I believe, although I'm not sure if it's only because then it can support `transferTo`, so this might be better

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    if (outputFileStream != null) {
+      outputFileStream.flush();
```
hmm should closing the `outputBufferedFileStream` close this stream as well?

ifilonenko replied:
Good point

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,171 @@
+  private void initStream() throws IOException {
+    if (outputFileStream == null) {
+      outputFileStream = new FileOutputStream(outputTempFile, true);
+      ts = new TimeTrackingOutputStream(metrics, outputFileStream);
```
why do you need a reference to `outputFileStream` and `ts`? It doesn't seem to be used for any other purpose than for constructing the `BufferedOutputStream`? I could be missing something though.

mccheah replied:
Shouldn't we always be writing to `ts` so that we get the time tracking? But this gets potentially inconsistent when considering trying to write to the output channels.

ifilonenko replied:
By writing to `outputBufferedFileStream` you get the benefits of `ts`. However, as you can see from the channel implementation. You are unable to use the this for `transferTo.` But its better to use `ts` where we can, imo.

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  @Override
+  public void commitAllPartitions() throws IOException {
```
do you need to close the `outputBufferedFileStream` stream in this method?

ifilonenko replied:
yes :) hehe

ifilonenko committed:
first pass of comments

mccheah commented on core/src/test/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriterSuite.scala:
```diff
@@ -118,6 +124,26 @@ class BypassMergeSortShuffleWriterSuite extends SparkFunSuite with BeforeAndAfte
         }
     })
+    when(shuffleExecutorComponents.writes()).thenReturn(writeSupport)
+    when(writeSupport.createMapOutputWriter(
```
Why not just use a real `DefaultShuffleWriteSupport`? This is an integration test so we can feel free to use real objects if they're on the critical path.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:
```diff
@@ -17,12 +17,10 @@
-import java.io.FileInputStream;
-import java.io.FileOutputStream;
-import java.io.IOException;
+import java.io.*;
```
Don't use * import

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:
```diff
@@ -103,70 +105,76 @@
       SparkConf conf,
-      ShuffleWriteMetricsReporter writeMetrics) {
+      ShuffleWriteMetricsReporter writeMetrics,
+      ShuffleExecutorComponents shuffleExecutorComponents) {
```
I think you can pass in just the `ShuffleWriteSupport` because you won't need the reader components here.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,196 @@
+  public void commitAllPartitions() throws IOException {
+    cleanUp();
+    blockResolver.writeIndexFileAndCommit(shuffleId, mapId, partitionLengths, outputTempFile);
+    if (!successfulWrite) {
```
There's a number of ways this can be done, right? For instance, can't we check that the file doesn't exist, and create it? That might be easier to read than using a boolean flag.

ifilonenko replied:
That happens later? its easier to ensure that some form of write happened by either the byte stream or byte channel as an initial filter, I would think

mccheah replied:
The action of committing would create the file if any bytes were written. But if no bytes are written, there would be no output file, right? Would be good to verify these behaviors by running `BypassMergeSortShuffleWriterSuite` locally and seeing what happens.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,196 @@
+
+    @Override
+    public WritableByteChannel openChannel() throws IOException {
+      return Channels.newChannel(outputFileStream);
```
This is strange - we create the `newChannel` around the `outputFileStream`, but `openStream` returns the  `DefaultShuffleBlockOutputStream`. This can lead to an inconsistency in the count returned by this writer, because if we return the `FileChannel` here, the counter from the `DefaultShuffleBlockOutputStream` will not be reflected.

I think most of this can be caught by running `BypassMergeSortShuffleWriterSuite` locally.

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:
```diff
@@ -179,37 +187,38 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
+          if (file.exists()) {
+            FileInputStream in = new FileInputStream(file);
+            try {
+              Utils.copyStream(in, tempOutputStream, false, false);
```
So using `Utils.copyStream` actually assumes that the outputStream is a `FileOutputStream` in order to take advantage of the `transferTo` logic inside that function. The function also does its own bufferring with an array of a fixed size. I wonder if we should just have the `DefaultShuffleMapOutputWriter` just deal with generic `FileOutputStream`s and wrap them with buffered output writers if necessary? It would make the `DefaultShuffleMapOutputWriter` much cleaner too so we're not keeping track of so many layers of outputStreams

yifeih replied:
Actually, on second thought, you probably just want to fork the code paths here into a "transferTo" block and a "outputstreams" block to allow more flexibility in the plugins to return whatever outputstreams they want.

yifeih commented on core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala:
```diff
@@ -127,6 +138,7 @@ private[spark] class SortShuffleManager(conf: SparkConf) extends ShuffleManager
       mapId: Int,
       context: TaskContext,
       metrics: ShuffleWriteMetricsReporter): ShuffleWriter[K, V] = {
+    initializeExecutor()
```
You dont' want to to do this at every `getWriter` call, but only once per executor. Since there's only one `SortShuffleManager` per executor, I think you can do it as a private variable in this class. You can get rid of the `initialized` flag then too.

ifilonenko replied:
Agreed

ifilonenko committed:
resolve some comments

ifilonenko committed:
small style choice with java.io.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,249 @@
+    public long getLength() {
+      if (channel != null && stream == null) {
+        try {
+          channel.close();
```
This will close the underlying file before we're done writing to it for subsequent partitions. This should be verifiable by running a unit test - `BypassMergeSortShuffleWriterSuite` - using multiple partitions.

mccheah replied:
Actually never mind sorry - just noticed that `channel` isn't the file channel but it's the shielded from close output channel.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,249 @@
+  private class DefaultShufflePartitionWriter implements ShufflePartitionWriter {
+
+    private final int partitionId;
+    private DefaultShuffleBlockOutputStream stream = null;
```
We might be able to rename these to something a bit clearer for their purpose - maybe `partitionWriterStream` and `partitionWriterChannel`. While it's more verbose, it makes it clear that these are only for writing the given partition - given that there's outside context that writes to the map output file. I got confused in https://github.com/bloomberg/apache-spark-on-k8s/pull/6/files#r268301953 because of the naming of these vars.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,249 @@
+        try {
+          channel.close();
+        } catch (Exception e) {
+          log.error("Error with closing channel for partition", e);
```
Shouldn't we throw if we fail to close here?

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,249 @@
+        try {
+          stream.close();
+        } catch (Exception e) {
+          log.error("Error with closing stream for partition", e);
```
Similarly might want to consider throwing here (need to make sure that interrupting the control flow retains correctness though).

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    if (outputFileStream != null) {
+      outputFileStream.close();
```
Does this close the file channel as well?

yifeih replied:
Yea, looking at the implementation of `FileOutputStream.close()`, calling close on the outputFileStream will also close the outputFileChannel

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,249 @@
+    if (!outputFile.delete() && outputFile.exists()) {
+      log.warn("Failed to delete outputshuffle file at {}", outputFile.getAbsolutePath());
+    }
+    cleanUp();
```
`cleanup` will close streams but the files will be deleted - will that throw an exception? I think what we want is something like this:

```
try {
  cleanup();
} catch (IOException e) {
  logger.error(...., e);
}
deleteFiles();
```

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,249 @@
+
+  private void cleanUp() throws IOException {
+    if (outputBufferedFileStream != null) {
+      outputBufferedFileStream.close();
```
Closing the buffered output stream will close the underlying file stream. Is closing the same file output stream twice prone to throwing an error? If it doesn't throw an error than this is fine (would rather be explicit about all the resources we're closing).

ifilonenko committed:
resolved more comments

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:
```diff
@@ -103,70 +110,76 @@
-      partitionWriters[partitioner.getPartition(key)].write(key, record._2());
-    }
+    ShuffleMapOutputWriter mapOutputWriter = shuffleWriteSupport
+      .createMapOutputWriter(appId, shuffleId, mapId, numPartitions);
```
quick question: can the `appId` actually just be passed to the `ShuffleMapOutputWriter` through the `ShuffleDataIO`? It should be part of the sparkConf and shouldn't change in the executors right?

ifilonenko replied:
It can, but its doesn't add much computation besides the above call to `getAppId()` so it seems pretty unintrusive.
However, the API was built so that you call:
```
  public ShuffleMapOutputWriter createMapOutputWriter(
      String appId,
      int shuffleId,
      int mapId,
      int numPartitions)
```
so I am a bit bounded by that :)

yifeih replied:
Yea I'm just wondering whether we need that in the API or not, since some implementations, like this refactor one that we're doing, don't necessarily need it, although all remote implementations might. @mccheah thoughts?

mccheah replied:
Yeah we can change the API, it should be passed through `ShuffleDataIO` - maybe `ShuffleExecutorComponents#initialize`?

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,250 @@
+            "Failed to create shuffle file directory at %s.",
+            outputFile.getParentFile().getAbsolutePath()));
+      }
+      if (!outputFile.isFile() && !outputFile.createNewFile()) {
```
Wait, i'm confused by this block. You're checking whether the output file is there, and then trying to create it? Shouldn't you be throwing if it's not there regardless of whether you can create it or not because when `blockResolver.writeIndexFileAndCommit()` is done, the file should be there? But let me know if I'm misunderstanding this.

mccheah replied:
If the committer didn't create the file, we want to try to create an empty one by default, I think. But it's strange to me that shuffles can write empty files. Was this the case before our plugin refactor?

yifeih replied:
Oh i see... well, looking at this code, I don't think it actually writes an empty file if there are no records: https://github.com/bloomberg/apache-spark-on-k8s/pull/6/files#diff-8b6b7a5dadc0d8e97307d0f8e8378d8fL126

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,250 @@
+    }
+  }
+
+  private void cleanUp() throws IOException {
```
You're closing the streams, but I noticed that `Channel` also has a `close()` method. Do you need to call `close()` on the `outputFileChannel` too?

yifeih replied:
Heh I am wrong... closing the fileOutputStream closes the channel :P

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,250 @@
+    }
+
+    @Override
+    public void close() {
```
I wonder if we need some equivalent of `flush()` here, especially since the method we use to write to the `WritableByteChannel` is called `Utils.copyFileStreamNIO`, which is non-blocking, so I'm not sure we're guaranteed to have written everything by the time `getCount()` is called. I was trying to do some research on byteChannels and the closest thing I found was `force()`

ifilonenko replied:
@mccheah any ideas here, I am a bit worried about this being non-blocking as well

mccheah replied:
I looked at `Utils.copyFileStreamNIO` which doesn't have any indication that it's non-blocking, any idea why we think it is?

yifeih replied:
I assumed `NIO` stood for non-blocking IO. The javadoc for the underlying method that it's using, `FileChannel.transferTo()` also implies that the underlying channel could be non-blocking. https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#transferTo(long,%20long,%20java.nio.channels.WritableByteChannel)

yifeih replied:
@mccheah and I just discussed some of this offline.

The `FileChannel.transferTo()` implementation checks for whether the `WriteableByteChannel` is an instance of `FileChannelImpl` and behaves different if it isn't. Because we're not passing in a `FileChannelImpl`, this might actually behave in ways that we don't expect that deviate from the current codepath. Therefore, we shouldn't be wrapping the the `FileChannel` with another `WritableByteChannel` as we're currently doing.

This limits our choices in how we can get the lengths to return in `DefaultShufflePartitionWriter.getLengths()`. The simplest way seems to be to use `FileChannel.position()` calls to get the positions when instantiating the byte channel and when closing the byte channel in `DefaultShufflePartitionWriter.getLengths()`.

Also, looking at the `DefaultShufflePartitionWriter.getLengths()` method, since we want to ensure that the shuffle partitions writers are all closed after getting the lengths, perhaps we should rename this method to `close()` so that it's clear what's should be happening (i.e. you can only get the lengths once everything is written)

yifeih commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -337,33 +337,16 @@ private[spark] object Utils extends Logging {
-         |after transferTo, please check your kernel version to see if it is 2.6.32,
-         |this is a kernel bug which will lead to unexpected behavior when using transferTo.
-         |You can set spark.file.transferTo = false to disable this NIO feature.
-           """.stripMargin)
```
read the links, and it seems to be a kernel bug? is this safe to delete?

ifilonenko replied:
There is no position in writeableByteChannels :/ soo... idk ... meep

mccheah replied:
If we keep this check it should be in the implementation of `ShuffleMapOutputWriter` I would guess. Maybe in `close` / `abort` / `commit` etc?

yifeih replied:
Perhaps i'm missing something, but how do you get the expected position in the `ShuffleMapOutputWriter` for the assert? If that's not easy to get, I would say it's better to keep this here with a if-statement to check if it's an instance of `FileChannel` to maintain the same behavior as before.

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleExecutorComponents.java:
```diff
@@ -0,0 +1,57 @@
+  public void intitializeExecutor(String appId, String execId) {
+    blockManager = SparkEnv.get().blockManager();
+    blockResolver = new IndexShuffleBlockResolver(sparkConf, blockManager);
+    metrics = TaskContext.get().taskMetrics();
```
wait, i don't think you want this here actually. The TaskContext is associated with each individual shuffle task, but you're only calling `initializeExecutor` once per executor. You want to get the metrics from `TaskContext` after you have the set the taskContext here (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/Task.scala#L102) for each task

ifilonenko replied:
@mccheah if so, can we pass `metrics` into the ShuffleMapOutputWriter so that it can be properly mocked (can't mock Static methods).

mccheah replied:
There's a few ways you can do it. You can pass it in as you suggested. In this particular case, you can also call `TaskContext.set(mockTaskContext)` to initialize the static variable before the test. I prefer the latter way, but you could make a case for either.

mccheah replied:
Actually you might want to pass it through so that concurrent tests don't collide...

mccheah replied:
The problem is that `TaskMetrics` is currently marked as `DeveloperApi` which means it's questionable to pass it in to a public API. We could proposed some alternative metrics API that delegates to the Spark default `TaskMetrics` API. But I think for now we can use `TaskContext#get` from inside the writer and then in tests call `TaskContext#setTaskContext`. I took a closer look and a lot of tests use `TaskContext#setTaskContext` indicating they don't anticipate tests to be run in parallel in the same JVM.

mccheah replied:
Just make sure we call `TaskContext#unset` appropriate after each test.

yifeih reviewed (commented):
overall, looks great! left a few comments and questions

ifilonenko committed:
initial unit tests of outputstream writing

ifilonenko committed:
slight update

ifilonenko committed:
added tests and resolved comments

ifilonenko committed:
initial comments resolved

ifilonenko committed:
trying to change API to return FileChannel instead of WriteableByteChannel

mccheah commented on core/src/main/java/org/apache/spark/api/shuffle/ShufflePartitionWriter.java:
```diff
@@ -34,9 +33,7 @@
-  long getLength();
+  long closeAndGetLength();
 
-  default WritableByteChannel openChannel() throws IOException {
```
Ah we talked offline but I still think this default applies.

mccheah commented on core/src/main/java/org/apache/spark/api/shuffle/ShufflePartitionWriter.java:
```diff
@@ -34,9 +33,7 @@
-  default WritableByteChannel openChannel() throws IOException {
-    return Channels.newChannel(openStream());
-  }
+  FileChannel openChannel() throws IOException;
```
This should still return `WritableByteChannel`, but the implementation returns `FileChannel` in `DefaultShufflePartitionWriter`.

mccheah replied:
The code in `Utils#transferTo` will eventually check `instanceOf` this `WritableByteChannel` and optimize on `FileChannel` accordingly,

ifilonenko committed:
small comments to resolve

ifilonenko committed:
fix to issue of closing channel

mccheah committed:
Various changes to the API and state assumptions in writers. (#8)

Proposes the following changes to the API:

- closeAndGetLength() is split into separate close() and getNumBytesWritten() operations.
- openChannel and openStream renamed to toChannel and toStream

Proposes the following changes to the implementation:

- close() in the default implementation now persists the length in the partitionLengths array
- getNumBytesWritten() doesn't necessitate the writer's resources to be closed ahead of it
- Don't close the stream in BypassMergeSortShuffleWriter - only close it in DefaultShufflePartitionWriter#close (for consistency with how we treat channels)

ifilonenko committed:
small comment addiiton

ifilonenko commented:
@yifeih for +1 :)

yifeih commented on core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:
```diff
@@ -179,37 +190,54 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
-            Closeables.close(in, copyThrewException);
+        boolean copyThrewException = true;
+        ShufflePartitionWriter writer = null;
+        try {
```
nit: i think you can reduce the amount of repeated code and nesting by moving the check for `file.exists()` to right after line 203
```
final File file = partitionWriterSegments[i].file();
if (!file.exists()) continue
boolean copyThrewException = true
ShufflePartitionWriter writer = null
try {
  writer = mapOutputWriter.getNextPartitionWriter();
  try {
    FileInputStream in = new FileInputStream(file)
    if (transferToEnabled) {
      WritableByteChannel outputChannel = writer.toChannel();
      FileChannel inputChannel = in.getChannel()
      Utils.copyFileStreamNIO(inputChannel, outputChannel, 0, inputChannel.size());
      copyThrewException = false;
    } else {
      OutputStream tempOutputStream = writer.toStream();
      Utils.copyStream(in, tempOutputStream, false, false);
      copyThrewException = false;
    }
    
  } finally {
    Closeables.close(in, copyThrewException);
  }
  if (!file.delete()) {
    logger.error();
  }
} finally {
  Closeables.close(writer, copyThrewException);
}

yifeih commented on core/src/main/java/org/apache/spark/api/shuffle/ShufflePartitionWriter.java:
```diff
@@ -31,12 +32,43 @@
+   * <p>
+   * Note that the default version of {@link #toChannel()} returns a {@link WritableByteChannel}
+   * that does not itself need to be closed up front; only the underlying output stream given by
+   * {@link #toStream()} must be closed.
```
I'm looking at the `Channels.newChannel()` implementation, and it uses `ReadableByteChannelImpl`, which all it seems to do is create an in-memory buffer, but it'll get garbage collected when the `WritableByteChannel` falls out of scope, so I think this is fine to not close that channel for now. However, if the `Channels.newChannel()` implementation changes to require additional resources, then we're out of luck, but that might not even be a valid concern?

mccheah replied:
We can cross that bridge when we get there - might be something to check for future versions of Java.

yifeih reviewed (commented):
looks great! just had a few small thoughts that might or might not be valid

ifilonenko committed:
resolving comments

yifeih commented:
@ifilonenko I dug into the making of the temp file. If you look at https://github.com/apache/spark/blob/master/core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java#L126, it's actually calling in with the `dataTmp` file set to null if there's nothing to iterate. This means that `writeIndexFileAndCommit()` is taking care of creating the temp file and whatever else it does. 

To preserve functionality as closely as possibly (in case there are other side effects down that code path), it's probably better to set `this.outputTempFile = null` in the constructor and create it in `getNextPartitionWriter()` if it's null.

ifilonenko committed:
resolving comments

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,217 @@
+    } catch (Exception e) {
+      log.error("Unable to close appropriate underlying file stream", e);
+    }
+    if (!outputTempFile.delete() && outputTempFile.exists()) {
```
Check for null

mccheah commented:
+1 from me after https://github.com/bloomberg/apache-spark-on-k8s/pull/6#discussion_r269338448

ifilonenko committed:
added null check

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,219 @@
+      if (!outputTempFile.delete() && outputTempFile.exists()) {
+        log.warn("Failed to delete temporary shuffle file at {}", outputTempFile.getAbsolutePath());
+      }
+      if (!outputFile.delete() && outputFile.exists()) {
```
Actually I don't think we want to delete here - a concurrent attempt shouldn't have its results removed if this attempt fails. I think this lines up with the behavior from the original shuffle writer.

yifeih commented:
looks good from my end! thanks for working on this!

mccheah commented on core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java:
```diff
@@ -0,0 +1,219 @@
+      log.error("Unable to close appropriate underlying file stream", e);
+    }
+    if (outputTempFile != null) {
+      if (!outputTempFile.delete() && outputTempFile.exists()) {
```
Switch the ordering of the statements in this boolean condition.
